### DISCUSSION TITLE
Exception at runtime when invoke service method, only happens when Service has a constructor

### DISCUSSION THREAD

**Discussion by mizzPpY** [CoreWCF/CoreWCF#974] [Q&A]
Hi all, 
I'm currently migrating .NET Framework 4.7.2 WCF project to .NET Core and encounter some issue that I couldn't understand.
When I run my project, and add the service URL in WCFTestClient.exe, the operations are listed out and it seems fine; but when I try to invoke any of them, will get an exception as following:
```
Exception caught at 
Method: InvokeAsync
Message:The service implementation object was not initialized or is not available.

Server stack trace: 
   at System.ServiceModel.Channels.ServiceChannel.HandleReply(ProxyOperationRuntime operation, ProxyRpc& rpc)
   at System.ServiceModel.Channels.ServiceChannel.Call(String action, Boolean oneway, ProxyOperationRuntime operation, Object[] ins, Object[] outs, TimeSpan timeout)
   at System.ServiceModel.Channels.ServiceChannelProxy.InvokeService(IMethodCallMessage methodCall, ProxyOperationRuntime operation)
   at System.ServiceModel.Channels.ServiceChannelProxy.Invoke(IMessage message)
```
I tried with both CoreWCF package **1.2.1** and **1.3.1**, the issue still persists. 
Here is my csproj :
```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="CoreWCF.ConfigurationManager" Version="1.2.1" />
  </ItemGroup>

  <ItemGroup>
    <!--ProjectReference  -->
  </ItemGroup>

</Project>
```

Here is my Program.cs
```
using CoreWCF;
using CoreWCF.Configuration;
using CoreWCF.Description;
using TestWebService.WebServices;

namespace TestWebService
{
    public class Program
    {        
        public static void Main (string[] args)
        {
            var builder = WebApplication.CreateBuilder(args);

            // Add CoreWCF services to the ASP.NET Core app's DI container
            builder.Services.AddServiceModelServices()
                            .AddServiceModelMetadata();
            
            var app = builder.Build();

            // Configure the HTTP request pipeline.
            if (!app.Environment.IsDevelopment())
            {
                app.UseExceptionHandler("/Error");
                // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
                app.UseHsts();
            }          

            // Configure CoreWCF endpoints in the ASP.NET Core host
            app.UseServiceModel(serviceBuilder =>
            {
                serviceBuilder.AddService<TestService>(options => { })
                .AddServiceEndpoint<TestService, ITestService>(new BasicHttpBinding(), "/TestService");
            });

            // Enable getting metadata/wsdl
            var serviceMetadataBehavior = app.Services.GetRequiredService<ServiceMetadataBehavior>();
            serviceMetadataBehavior.HttpGetEnabled = true;
            serviceMetadataBehavior.MetadataExporter.PolicyVersion = PolicyVersion.Default;

            app.Run();
        }        
    }
}
```

Here is TestService.cs:
```
using Common;
using Common.Enums;
using CoreWCF;
using DataModelPoco.Models;
using ErrorHandler;

namespace TestWebService.WebServices
{
    [ServiceBehavior(InstanceContextMode = InstanceContextMode.PerSession, ConcurrencyMode = ConcurrencyMode.Multiple, IncludeExceptionDetailInFaults = true)]
    [GlobalErrorBehavior(typeof(GlobalErrorHandler))]
    public class TestService : ITestService
    {
        private string placeHolderParam;
        private int placeHolderParam2;

        public TestService(string param1, int param2)
        {
            placeHolderParam = param1;
            placeHolderParam2 = param2;
        }

        public string GetAString(string random)
        {
            return $"Finally return the string {random} !";
        }  
    }
}
```

Later I found out, if I removed the constructor from TestService, the operation can be called successfully from WCFTestClient.exe. 
How is the constructor "blocking" the service from working (the constructor show above is just simple sample with simple int and string type params, in real case the constructor will be accepting complex type as param)? Am I missing something in my Program.cs?

**Comment by g7ed6e** [comment]
@mizzPpY You need to add `TestService` to the ServiceCollection of the WebApplication builder if you want CoreWCF to pull an instance from the DI container.
```
builder.Services.AddTransient<TestService>();
```

**Reply by mizzPpY** [reply]
Thank you for your prompt reply @g7ed6e !

I tried the following modification in Program.cs.
Attempt no.1 - **builder.Services.AddTransient<TestService>()**:
```
public static void Main (string[] args)
        {
            var builder = WebApplication.CreateBuilder(args);

            builder.Services.AddTransient<TestService>();     //Add service to service collection       
            
            // Add CoreWCF services to the ASP.NET Core app's DI container
            builder.Services.AddServiceModelServices()
                            .AddServiceModelMetadata();            
            
            var app = builder.Build();      //throw exception upon starting the service

            // Configure the HTTP request pipeline.
            if (!app.Environment.IsDevelopment())
            {
                app.UseExceptionHandler("/Error");
                // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
                app.UseHsts();
            }            

            // Configure CoreWCF endpoints in the ASP.NET Core host
            app.UseServiceModel(serviceBuilder =>
            {
                serviceBuilder.AddService<TestService>(options => { })
                .AddServiceEndpoint<TestService, ITestService>(new BasicHttpBinding(), "/TestService");
            });

            // Enable getting metadata/wsdl
            var serviceMetadataBehavior = app.Services.GetRequiredService<ServiceMetadataBehavior>();
            serviceMetadataBehavior.HttpGetEnabled = true;
            serviceMetadataBehavior.MetadataExporter.PolicyVersion = PolicyVersion.Default;

            app.Run();
        }
```
Exception thrown upon starting the service at line "builder.Build()"

> System.AggregateException: 'Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: TestWebService.WebServices.TestService Lifetime: Transient ImplementationType: TestWebService.WebServices.TestService': Unable to resolve service for type 'System.String' while attempting to activate 'TestWebService.WebServices.TestService'.)'

Attempt no.2 - **builder.Services.<ITestService, TestService>(x => new TestService("", 0))**:
```
public static void Main (string[] args)
        {
            var builder = WebApplication.CreateBuilder(args);

            builder.Services.<ITestService, TestService>(x => new TestService("", 0));     //Add service to service collection       
            
            // Add CoreWCF services to the ASP.NET Core app's DI container
            builder.Services.AddServiceModelServices()
                            .AddServiceModelMetadata();            
            
            var app = builder.Build(); 

            // Configure the HTTP request pipeline.
            if (!app.Environment.IsDevelopment())
            {
                app.UseExceptionHandler("/Error");
                // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
                app.UseHsts();
            }            

            // Configure CoreWCF endpoints in the ASP.NET Core host
            app.UseServiceModel(serviceBuilder =>
            {
                serviceBuilder.AddService<TestService>(options => { })
                .AddServiceEndpoint<TestService, ITestService>(new BasicHttpBinding(), "/TestService");
            });

            // Enable getting metadata/wsdl
            var serviceMetadataBehavior = app.Services.GetRequiredService<ServiceMetadataBehavior>();
            serviceMetadataBehavior.HttpGetEnabled = true;
            serviceMetadataBehavior.MetadataExporter.PolicyVersion = PolicyVersion.Default;

            app.Run();
        }
```
This successfully compile and start up, however when trigger an operation from TestService, gets a similar error with the original issue above:

> The service implementation object was not initialized or is not available.
> 
> Server stack trace: 
>    at System.ServiceModel.Channels.ServiceChannel.ThrowIfFaultUnderstood(Message reply, MessageFault fault, String action, MessageVersion version, FaultConverter faultConverter)
>    at System.ServiceModel.Channels.ServiceChannel.HandleReply(ProxyOperationRuntime operation, ProxyRpc& rpc)
>    at System.ServiceModel.Channels.ServiceChannel.Call(String action, Boolean oneway, ProxyOperationRuntime operation, Object[] ins, Object[] outs, TimeSpan timeout)
>    at System.ServiceModel.Channels.ServiceChannelProxy.InvokeService(IMethodCallMessage methodCall, ProxyOperationRuntime operation)
>    at System.ServiceModel.Channels.ServiceChannelProxy.Invoke(IMessage message)

**Reply by mconnew** [reply]
The problem with attempt 1 is because you can't really inject a string and an int into DI. If the TestService constructor was declared like this:
```c#
public TestService(ILogger<TestService> logger)
{
    _logger = logger;
}
```
Then as `ILogger<TestService>` is obtainable from DI, it can be provided to the constructor. The existing sample as is can't pull a string and an int from DI. I suspect you can probably inject one of each into DI, but it would be a bit strange to say `builder.Services.AddTransient<string>("");` or `builder.Services.AddTransient<int>(0);`.  

The problem with the second attempt is in the overload being used (and the missing AddTransient method name, but I suspect that was a typo from trying different things). You are adding a type of `ITestService` to DI, but CoreWCF is looking for `TestService`. Change it to this and it will work:
```
builder.Services.AddTransient<TestService>(x => new TestService("", 0));     //Add service to service collection       
```
Off the top of my head, I can't remember if there's a generic overload which only takes a type which applies to both the service type and the concrete type when using a factory delegate, so it's possible you might need the follow (but I think the previous suggestion should work):
```
builder.Services.AddTransient<TestService,TestService>(x => new TestService("", 0));     //Add service to service collection       
```

**Reply by mizzPpY** [reply]
Thank you @mconnew for your explanation, I didn't know about the following

> The problem with attempt 1 is because you can't really inject a string and an int into DI.

In my actual project I will be passing in custom classes into TestService, but that didn't work in the beginning, so I thought maybe I take a step back and start with simple type like string and int.

With your suggestion, it is finally starting and able to call method in TestService with
   ```
   builder.Services.AddTransient<TestService>(x => new TestService("", 0)); //WORKS! 
            
   // Add CoreWCF services to the ASP.NET Core app's DI container
   builder.Services.AddServiceModelServices()
                  .AddServiceModelMetadata();
```         

I also tested with passing custom class object into TestService.
TestService.cs
```
private CustomClass _customObject;
public TestService(CustomClass customObject)
{
    _customObject = customObject;
}
```

Program.cs:

```
public static CustomClass TheCustomObject;
        
        public static void Main (string[] args)
        {
            var builder = WebApplication.CreateBuilder(args);

            TheCustomObject = new CustomClass(builder.Configuration); //should this be builder.Services.AddTransient<>()?

            builder.Services.AddTransient<TestService>(x => new TestService(TheCustomObject)); //WORKS too! 
            
            // Add CoreWCF services to the ASP.NET Core app's DI container
            builder.Services.AddServiceModelServices()
                            .AddServiceModelMetadata();
                                    
            var app = builder.Build();

            .
            .
            . 
            app.Run();
        }
```
For this case, is this acceptable (because is running fine), or I must first register CustomClass object in DI before add TestService? 
`builder.Services.AddTransient<CustomClass>(x => new CustomClass(builder.Configuration));`

**Comment by g7ed6e** [comment]
@mizzPpY Attempt 1 is correct. It looks like the constructor injected dependencies of your `TestService` are not registered..
> System.AggregateException: 'Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: TestWebService.WebServices.TestService Lifetime: Transient ImplementationType: TestWebService.WebServices.TestService': Unable to resolve service for type 'System.String' while attempting to activate 'TestWebService.WebServices.TestService'.)'
